File: lib/interop/stream_writer.rb
require 'interop/message'
require 'interop/writer'
require 'interop/stream_adapter'

module Hx
  module Interop
    # Writes messages to a stream (e.g. STDOUT)
    class StreamWriter < StreamAdapter
      include Writer

      protected

      # @param [Message] message
      def _write(message)
        @mutex.synchronize do
          message.headers.each do |k, v|
            @stream.puts "#{k}: #{v}"
          end
          @stream.puts
          @stream.write message.body
          @stream.puts
        end
      end
    end
  end
end

File: php/Interop/StreamWriter.php
<?php

namespace Hx\Interop;

class StreamWriter implements Writer {
    private ByteWriter $byteWriter;

    /**
     * @param ByteWriter $byteWriter
     */
    public function __construct(ByteWriter $byteWriter) {
        $this->byteWriter = $byteWriter;
    }

    public function write(Message $message): void {
        foreach($message->headers as $header) {
            $this->byteWriter->writeBytes(strval($header));
        }
        $this->byteWriter->writeBytes("\n");
        $this->byteWriter->writeBytes($message->body);
        $this->byteWriter->writeBytes("\n");
    }
}
